arch/x86/core/userspace.S

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <arch/x86/asm.h>
+#include <offsets_short.h>
+
+GTEXT(_x86_syscall_invoke)
+GTEXT(_x86_syscall_entry_stub)
+GTEXT(_x86_syscall_entry)
+GTEXT(_x86_userspace_enter)
+
+/* Landing site for syscall SW IRQ. Marshal arguments and call C function for
+ * further processing.
+ */
+SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
+	push	%esi		/* arg5 */
+	push	%edi		/* arg4 */
+	push	%ebx		/* arg3 */
```

Comment:
Sort of annoying that you need to push stuff on the stack for arguments which are only going to be valid for a small minority of syscalls.  Maybe more cycle-counting than it's worth, but it wouldn't be *that* much more work to have a separate syscall interrupt registered for each argument count.

Reply:
gonna punt on this for now, but maybe revisit in the future. seems like other OSes use a single vector for system calls.